File: src/test/java/com/lehvolk/xodus/web/services/SmartSearchQueryParserTest.java
package com.lehvolk.xodus.web.services;

import java.util.Collections;
import java.util.List;

import org.junit.Test;

import com.lehvolk.xodus.web.services.SearchTerm.Range;
import com.lehvolk.xodus.web.services.SearchTerm.SearchTermType;
import static org.junit.Assert.assertEquals;

public class SmartSearchQueryParserTest {

    private static final String FIRST_NAME = "firstName";
    private static final String LAST_NAME = "lastName";

    @Test
    public void testSimple() {
        List<SearchTerm<?>> result = parse("firstName='John' and lastName='McClane'");
        assertEquals(2, result.size());
        verifyTerm(result.get(0), FIRST_NAME, "John", SearchTermType.VALUE);
        verifyTerm(result.get(1), LAST_NAME, "McClane", SearchTermType.VALUE);
    }

    @Test
    public void testSingleParam() {
        List<SearchTerm<?>> result = parse("firstName='John'");
        assertEquals(1, result.size());
        verifyTerm(result.get(0), FIRST_NAME, "John", SearchTermType.VALUE);
    }

    @Test
    public void testWithoutQuotes() {
        List<SearchTerm<?>> result = parse("firstName=John and lastName=McClane and age=43");
        assertEquals(3, result.size());
        verifyTerm(result.get(0), FIRST_NAME, "John", SearchTermType.VALUE);
        verifyTerm(result.get(1), LAST_NAME, "McClane", SearchTermType.VALUE);
        verifyTerm(result.get(2), "age", "43", SearchTermType.VALUE);
    }

    @Test
    public void testLike() {
        List<SearchTerm<?>> result = parse("firstName=John and lastName~McClane");
        assertEquals(2, result.size());
        verifyTerm(result.get(0), FIRST_NAME, "John", SearchTermType.VALUE);
        verifyTerm(result.get(1), LAST_NAME, "McClane", SearchTermType.LIKE);
    }

    @Test
    public void testSingleParamsWithoutQuotes() {
        List<SearchTerm<?>> result = parse("firstName=John");
        assertEquals(1, result.size());
        verifyTerm(result.get(0), FIRST_NAME, "John", SearchTermType.VALUE);
    }

    @Test
    public void testEscapingQuotes() {
        List<SearchTerm<?>> result = parse("firstName=John and lastName='Mc''Clane'");
        assertEquals(2, result.size());
        verifyTerm(result.get(0), FIRST_NAME, "John", SearchTermType.VALUE);
        verifyTerm(result.get(1), LAST_NAME, "Mc'Clane", SearchTermType.VALUE);
    }

    @Test
    public void testOmitAndIntoQuotes() {
        List<SearchTerm<?>> result = parse("firstName='John and Mike' and lastName='McClane'");
        assertEquals(2, result.size());
        verifyTerm(result.get(0), FIRST_NAME, "John and Mike", SearchTermType.VALUE);
        verifyTerm(result.get(1), LAST_NAME, "McClane", SearchTermType.VALUE);
    }

    @Test
    public void testEqualsIntoQuotes() {
        List<SearchTerm<?>> result = parse("firstName='John=Mike' and lastName='McClane'");
        assertEquals(2, result.size());
        verifyTerm(result.get(0), FIRST_NAME, "John=Mike", SearchTermType.VALUE);
        verifyTerm(result.get(1), LAST_NAME, "McClane", SearchTermType.VALUE);
    }

    @Test
    public void testSpacesIntoQuotes() {
        List<SearchTerm<?>> result = parse("firstName='John Mike' and lastName='McClane'");
        assertEquals(2, result.size());
        verifyTerm(result.get(0), FIRST_NAME, "John Mike", SearchTermType.VALUE);
        verifyTerm(result.get(1), LAST_NAME, "McClane", SearchTermType.VALUE);
    }

    @Test
    public void testSpecialSymbols() {
        List<SearchTerm<?>> result = parse("'_!@firstName'='John Mike' and '_!@lastName'='McClane'");
        assertEquals(2, result.size());
        verifyTerm(result.get(0), "_!@firstName", "John Mike", SearchTermType.VALUE);
        verifyTerm(result.get(1), "_!@lastName", "McClane", SearchTermType.VALUE);
    }

    @Test
    public void testRange() {
        List<SearchTerm<?>> result = parse("firstName='John Mike' and age=[30,40]");
        assertEquals(2, result.size());
        verifyTerm(result.get(0), FIRST_NAME, "John Mike", SearchTermType.VALUE);
        verifyRangeTerm(result.get(1), "age", 30, 40);
    }


    @Test(expected = ParseException.class)
    public void testCheckThrowsExceptionOnEmptyMap() throws ParseException {
        SmartSearchQueryParser.check(Collections.emptyList(), "");
    }

    @Test(expected = IllegalStateException.class)
    public void toolkitShouldThrowIllegalStateExceptionOnInit() throws ParseException {
        SmartSearchToolkit.newParser(null);
    }

    private static List<SearchTerm<?>> parse(String query) {
        try {
            return SmartSearchToolkit.newParser(query).parse();
        } catch (Exception e) {
            throw new IllegalStateException(e);
        }
    }

    @Test
    public void testCaseInsensitive() {
        List<SearchTerm<?>> result = parse("firstName='John Mike' AND lastName='McClane'");
        assertEquals(2, result.size());
        verifyTerm(result.get(0), FIRST_NAME, "John Mike", SearchTermType.VALUE);
        verifyTerm(result.get(1), LAST_NAME, "McClane", SearchTermType.VALUE);
    }

    private void verifyTerm(SearchTerm<?> term, String property, String value, SearchTermType type) {
        assertEquals(property, term.getProperty());
        assertEquals(value, term.getValue());
        assertEquals(type, term.getType());
    }

    @SuppressWarnings("unchecked")
    private void verifyRangeTerm(SearchTerm<?> term, String property, long start, long end) {
        assertEquals(SearchTermType.RANGE, term.getType());
        SearchTerm<Range> rangedTerm = (SearchTerm<Range>) term;
        assertEquals(property, term.getProperty());
        assertEquals(start, rangedTerm.getValue().getStart());
        assertEquals(end, rangedTerm.getValue().getEnd());
    }

}

File: entity-browser-app/src/test/kotlin/com/lehvolk/xodus/web/search/SmartSearchQueryParserTest.kt
package com.lehvolk.xodus.web.search


import org.junit.Assert.assertEquals
import org.junit.Test

class SmartSearchQueryParserTest {

    private val FIRST_NAME = "firstName"
    private val LAST_NAME = "lastName"


    @Test
    fun testSimple() {
        val result = "firstName='John' and lastName='McClane'".parse()
        assertEquals(2, result.size.toLong())
        verifyTerm(result[0], FIRST_NAME, "John", SearchTermType.VALUE)
        verifyTerm(result[1], LAST_NAME, "McClane", SearchTermType.VALUE)
    }

    @Test
    fun testSingleParam() {
        val result = "firstName='John'".parse()
        assertEquals(1, result.size.toLong())
        verifyTerm(result[0], FIRST_NAME, "John", SearchTermType.VALUE)
    }

    @Test
    fun testWithoutQuotes() {
        val result = "firstName=John and lastName=McClane and age=43".parse()
        assertEquals(3, result.size.toLong())
        verifyTerm(result[0], FIRST_NAME, "John", SearchTermType.VALUE)
        verifyTerm(result[1], LAST_NAME, "McClane", SearchTermType.VALUE)
        verifyTerm(result[2], "age", "43", SearchTermType.VALUE)
    }

    @Test
    fun testLike() {
        val result = "firstName=John and lastName~McClane".parse()
        assertEquals(2, result.size.toLong())
        verifyTerm(result[0], FIRST_NAME, "John", SearchTermType.VALUE)
        verifyTerm(result[1], LAST_NAME, "McClane", SearchTermType.LIKE)
    }

    @Test
    fun testSingleParamsWithoutQuotes() {
        val result = "firstName=John".parse()
        assertEquals(1, result.size.toLong())
        verifyTerm(result[0], FIRST_NAME, "John", SearchTermType.VALUE)
    }

    @Test
    fun testEscapingQuotes() {
        val result = "firstName=John and lastName='Mc''Clane'".parse()
        assertEquals(2, result.size.toLong())
        verifyTerm(result[0], FIRST_NAME, "John", SearchTermType.VALUE)
        verifyTerm(result[1], LAST_NAME, "Mc'Clane", SearchTermType.VALUE)
    }

    @Test
    fun testOmitAndIntoQuotes() {
        val result = "firstName='John and Mike' and lastName='McClane'".parse()
        assertEquals(2, result.size.toLong())
        verifyTerm(result[0], FIRST_NAME, "John and Mike", SearchTermType.VALUE)
        verifyTerm(result[1], LAST_NAME, "McClane", SearchTermType.VALUE)
    }

    @Test
    fun testEqualsIntoQuotes() {
        val result = "firstName='John=Mike' and lastName='McClane'".parse()
        assertEquals(2, result.size.toLong())
        verifyTerm(result[0], FIRST_NAME, "John=Mike", SearchTermType.VALUE)
        verifyTerm(result[1], LAST_NAME, "McClane", SearchTermType.VALUE)
    }

    @Test
    fun testSpacesIntoQuotes() {
        val result = "firstName='John Mike' and lastName='McClane'".parse()
        assertEquals(2, result.size.toLong())
        verifyTerm(result[0], FIRST_NAME, "John Mike", SearchTermType.VALUE)
        verifyTerm(result[1], LAST_NAME, "McClane", SearchTermType.VALUE)
    }

    @Test
    fun testSpecialSymbols() {
        val result = "'_!@firstName'='John Mike' and '_!@lastName'='McClane'".parse()
        assertEquals(2, result.size.toLong())
        verifyTerm(result[0], "_!@firstName", "John Mike", SearchTermType.VALUE)
        verifyTerm(result[1], "_!@lastName", "McClane", SearchTermType.VALUE)
    }

    @Test
    fun testRange() {
        val result = "firstName='John Mike' and age=[30,40]".parse()
        assertEquals(2, result.size.toLong())
        verifyTerm(result[0], FIRST_NAME, "John Mike", SearchTermType.VALUE)
        verifyRangeTerm(result[1], "age", 30, 40)
    }

    @Test
    fun testLink() {
        val result = "@linkName=MyType[3] and firstName~Jo".parse()
        assertEquals(2, result.size)
        verifyTerm(result[0], "@linkName", "MyType[3]", SearchTermType.VALUE, SearchType.LINK)
        verifyTerm(result[1], FIRST_NAME, "Jo", SearchTermType.LIKE)
    }

    @Test
    fun testPropNull() {
        val result = "firstName=null and lastName='null'".parse()
        assertEquals(2, result.size)
        verifyTerm(result[0], FIRST_NAME, null, SearchTermType.VALUE)
        verifyTerm(result[1], LAST_NAME, "null", SearchTermType.VALUE)
    }

    @Test
    fun testLinkNull() {
        val result = "@user=null and @action=MyAction[4]".parse()
        assertEquals(2, result.size)
        verifyTerm(result[0], "@user", null, SearchTermType.VALUE, SearchType.LINK)
        verifyTerm(result[1], "@action", "MyAction[4]", SearchTermType.VALUE, SearchType.LINK)
    }

    @Test(expected = ParseException::class)
    @Throws(ParseException::class)
    fun testCheckThrowsExceptionOnEmptyMap() {
        SmartSearchQueryParser.check(emptyList<SearchTerm>(), "")
    }

    @Test
    fun testCaseInsensitive() {
        val result = "firstName='John Mike' AND lastName='McClane'".parse()
        assertEquals(2, result.size.toLong())
        verifyTerm(result[0], FIRST_NAME, "John Mike", SearchTermType.VALUE)
        verifyTerm(result[1], LAST_NAME, "McClane", SearchTermType.VALUE)
    }

    private fun verifyTerm(term: SearchTerm, property: String, value: String?, type: SearchTermType, searchType: SearchType = SearchType.PROPERTY) {
        assertEquals(property, term.property)
        assertEquals(value, term.value)
        assertEquals(type, term.termType)
        assertEquals(searchType, term.searchType)
    }

    @SuppressWarnings("unchecked")
    private fun verifyRangeTerm(term: SearchTerm, property: String, start: Long, end: Long) {
        assertEquals(SearchTermType.RANGE, term.termType)
        val rangedTerm = term.value as SearchTerm.Range
        assertEquals(property, term.property)
        assertEquals(start, rangedTerm.start)
        assertEquals(end, rangedTerm.end)
    }

}
